File: mini-app/src/api/federations.ts
import api from './client'

export interface Federation {
  id: string
  name: string
  description: string | null
  is_public: boolean
  owner_id: number
  created_at: string
  owner: {
    id: number
    username: string | null
    first_name: string
  }
  member_count?: number
  admin_count?: number
}

export interface FederationBan {
  id: number
  target_user: {
    id: number
    username: string | null
    first_name: string
    last_name: string | null
  }
  banned_by: {
    id: number
    username: string | null
    first_name: string
  }
  reason: string
  created_at: string
  expires_at: string | null
}

export const createFederation = async (data: { name: string; description?: string; is_public?: boolean }) => {
  const response = await api.post('/federations', data)
  return response.data
}

export const getMyFederations = async () => {
  const response = await api.get('/federations/my')
  return response.data
}

export const getFederation = async (fedId: string) => {
  const response = await api.get(`/federations/${fedId}`)
  return response.data
}

export const updateFederation = async (fedId: string, data: { name?: string; description?: string }) => {
  const response = await api.patch(`/federations/${fedId}`, data)
  return response.data
}

export const deleteFederation = async (fedId: string) => {
  const response = await api.delete(`/federations/${fedId}`)
  return response.data
}

export const joinFederation = async (fedId: string) => {
  const response = await api.post(`/federations/${fedId}/join`)
  return response.data
}

export const leaveFederation = async (fedId: string) => {
  const response = await api.post(`/federations/${fedId}/leave`)
  return response.data
}

export const getFederationBans = async (fedId: string) => {
  const response = await api.get(`/federations/${fedId}/bans`)
  return response.data
}

export const federateBan = async (
  fedId: string,
  data: { user_id: number; reason: string; duration?: string }
) => {
  const response = await api.post(`/federations/${fedId}/ban`, data)
  return response.data
}

export const federateUnban = async (fedId: string, userId: number) => {
  const response = await api.delete(`/federations/${fedId}/ban/${userId}`)
  return response.data
}

export const getFederationChats = async (fedId: string) => {
  const response = await api.get(`/federations/${fedId}/chats`)
  return response.data
}

export const addFederationAdmin = async (fedId: string, userId: number) => {
  const response = await api.post(`/federations/${fedId}/admins`, { user_id: userId })
  return response.data
}

export const removeFederationAdmin = async (fedId: string, userId: number) => {
  const response = await api.delete(`/federations/${fedId}/admins/${userId}`)
  return response.data
}

export const getFederationAdmins = async (fedId: string) => {
  const response = await api.get(`/federations/${fedId}/admins`)
  return response.data
}

File: api/routers/federations.py
"""Federations router."""

from typing import List
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from api.routers.auth import get_current_user
from shared.database import get_db
from shared.models import (
    Federation,
    FederationAdmin,
    FederationBan,
    FederationMember,
    Group,
    Member,
    User,
)
from shared.schemas import (
    FederationCreate,
    FederationResponse,
    FederationUpdate,
)

router = APIRouter()


@router.get("/federations", response_model=List[FederationResponse])
async def list_federations(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List federations owned by current user."""
    result = await db.execute(
        select(Federation).where(Federation.owner_id == current_user.id)
    )
    return result.scalars().all()


@router.post("/federations", response_model=FederationResponse)
async def create_federation(
    request: FederationCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Create a new federation."""
    federation = Federation(
        name=request.name,
        description=request.description,
        owner_id=current_user.id,
        is_public=request.is_public,
    )
    db.add(federation)
    await db.flush()

    # Add owner as admin
    admin = FederationAdmin(
        federation_id=federation.id,
        user_id=current_user.id,
        added_by=current_user.id,
    )
    db.add(admin)
    await db.commit()

    return federation


@router.get("/federations/{fed_id}", response_model=FederationResponse)
async def get_federation(
    fed_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get federation details."""
    result = await db.execute(
        select(Federation).where(Federation.id == fed_id)
    )
    federation = result.scalar()

    if not federation:
        raise HTTPException(status_code=404, detail="Federation not found")

    return federation


@router.post("/federations/{fed_id}/join/{group_id}")
async def join_federation(
    fed_id: UUID,
    group_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Join a group to a federation."""
    # Check group ownership
    result = await db.execute(
        select(Group).where(
            Group.id == group_id,
            Group.owner_id == current_user.id,
        )
    )
    if not result.scalar():
        raise HTTPException(status_code=403, detail="Group ownership required")

    # Get federation
    result = await db.execute(
        select(Federation).where(Federation.id == fed_id)
    )
    if not result.scalar():
        raise HTTPException(status_code=404, detail="Federation not found")

    # Check if already a member
    result = await db.execute(
        select(FederationMember).where(
            FederationMember.federation_id == fed_id,
            FederationMember.group_id == group_id,
        )
    )
    if result.scalar():
        raise HTTPException(status_code=400, detail="Already a member")

    # Join
    member = FederationMember(
        federation_id=fed_id,
        group_id=group_id,
        joined_by=current_user.id,
    )
    db.add(member)
    await db.commit()

    return {"success": True}


@router.post("/federations/{fed_id}/ban/{user_id}")
async def fedban_user(
    fed_id: UUID,
    user_id: int,
    reason: str = "No reason provided",
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Ban a user from a federation."""
    # Check federation admin
    result = await db.execute(
        select(FederationAdmin).where(
            FederationAdmin.federation_id == fed_id,
            FederationAdmin.user_id == current_user.id,
        )
    )
    if not result.scalar():
        raise HTTPException(status_code=403, detail="Federation admin required")

    # Create ban
    ban = FederationBan(
        federation_id=fed_id,
        target_user_id=user_id,
        banned_by=current_user.id,
        reason=reason,
    )
    db.add(ban)
    await db.commit()

    return {"success": True}


@router.get("/federations/{fed_id}/bans")
async def list_fedbans(
    fed_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List federation bans."""
    result = await db.execute(
        select(FederationBan, User)
        .join(User, FederationBan.target_user_id == User.id)
        .where(FederationBan.federation_id == fed_id)
        .order_by(FederationBan.created_at.desc())
    )

    bans = [
        {
            "id": ban.id,
            "user": {
                "id": user.id,
                "username": user.username,
                "first_name": user.first_name,
            },
            "reason": ban.reason,
            "created_at": ban.created_at,
        }
        for ban, user in result.all()
    ]

    return bans
